python/lsst/ts/ess/common/device/base_device.py

```diff
@@ -0,0 +1,167 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BaseDevice"]
+
+from abc import ABC, abstractmethod
+import asyncio
+import logging
+from typing import Callable, List, Union
+
+from ..constants import Key, ResponseCode
+from ..sensor import BaseSensor
+from lsst.ts import utils
+
+
+class BaseDevice(ABC):
+    """Base class for the different types of Sensor Devices.
+
+    This class holds all common code for the hardware devices. Device specific
+    code (for instance for a serial or an FTDI device) needs to be implemented
+    in a sub-class.
+
+    Parameters
+    ----------
+    name: `str`
+        The name of the device.
+    device_id: `str`
+        The hardware device ID to connect to. This can be a physical ID (e.g.
+        /dev/ttyUSB0), a serial port (e.g. serial_ch_1) or any other ID used by
+        the specific device.
+    sensor: `BaseSensor`
+        The sensor that produces the telemetry.
+    callback_func : `Callable`
+        Callback function to receive the telemetry.
+    log: `logging.Logger`
+        The logger to create a child logger for.
+    """
+
+    @abstractmethod
+    def __init__(
+        self,
+        name: str,
+        device_id: str,
+        sensor: BaseSensor,
+        callback_func: Callable,
+        log: logging.Logger,
+    ) -> None:
+        self.name: str = name
+        self._device_id: str = device_id
+        self._sensor: BaseSensor = sensor
+        self._callback_func: Callable = callback_func
+        self._telemetry_loop: asyncio.Future = utils.make_done_future()
+        self.is_open = False
+        self.log = log.getChild(type(self).__name__)
+
+        # Support MockDevice fault state. To be used in unit tests only.
+        self._in_error_state: bool = False
+
+    async def open(self) -> None:
+        """Generic open function.
+
+        Check if the device is open and, if not, call basic_open. Then start
+        the telemetry loop.
+
+        Raises
+        ------
+        RuntimeError
+            In case the device already is open.
+        """
+        if self.is_open:
+            self.log.error("Already open, ignoring.")
+        await self.basic_open()
+        self.is_open = True
+
+        self.log.debug(f"Starting read loop for {self.name!r} sensor.")
+        self._telemetry_loop = asyncio.create_task(self._run())
+
+    @abstractmethod
+    async def basic_open(self) -> None:
+        """Open the Sensor Device."""
+        pass
+
+    async def _run(self) -> None:
+        """Run sensor read loop.
+
+        If enabled, loop and read the sensor and pass result to callback_func.
+        """
+        self.log.debug("Starting sensor.")
+        while not self._telemetry_loop.done():
+            self.log.debug("Reading data.")
+            curr_tai: float = utils.current_tai()
+            response: int = ResponseCode.OK
```

Comment:
I added the type annotations because that way my IDE was able to do code completion and without not. Is it a big issue if I left the type annotations there?